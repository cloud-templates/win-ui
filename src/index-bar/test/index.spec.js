import { mount, trigger, triggerDrag, mockScrollIntoView } from '../../../test';

function mockOffsetHeight(offsetHeight) {
  Object.defineProperty(HTMLElement.prototype, 'offsetHeight', {
    get() {
      return offsetHeight;
    },
  });
}

test('custom anchor text', () => {
  const wrapper = mount({
    template: `
      <win-index-bar>
        <win-index-anchor index="A">Title A</win-index-anchor>
        <win-index-anchor index="B">Title B</win-index-anchor>
      </win-index-bar>
    `,
  });

  expect(wrapper).toMatchSnapshot();
});

test('click and scroll to anchor', () => {
  const onSelect = jest.fn();
  const wrapper = mount({
    template: `
      <win-index-bar @select="onSelect">
        <win-index-anchor index="A" />
        <win-index-anchor index="B" />
      </win-index-bar>
    `,
    methods: {
      onSelect,
    },
  });

  const fn = mockScrollIntoView();
  const indexes = wrapper.findAll('.win-index-bar__index');
  indexes.at(0).trigger('click');

  expect(fn).toHaveBeenCalledTimes(1);
  expect(onSelect).toHaveBeenCalledWith('A');
});

test('touch and scroll to anchor', () => {
  const onSelect = jest.fn();
  const wrapper = mount({
    template: `
      <win-index-bar @select="onSelect">
        <win-index-anchor index="A" />
        <win-index-anchor index="B" />
        <win-index-anchor index="XXX" />
      </win-index-bar>
    `,
    methods: {
      onSelect,
    },
  });

  const fn = mockScrollIntoView();
  const sidebar = wrapper.find('.win-index-bar__sidebar');
  const indexes = wrapper.findAll('.win-index-bar__index');

  document.elementFromPoint = function (x, y) {
    const index = y / 100;

    if (index === 1 || index === 2) {
      return indexes.at(index).element;
    }

    if (index === 3) {
      return {
        dataset: {},
      };
    }
  };

  // horizontal drag
  triggerDrag(sidebar, 100, 0);
  expect(fn).toHaveBeenCalledTimes(0);

  // vertical drag
  trigger(sidebar, 'touchstart', 0, 0);
  trigger(sidebar, 'touchmove', 0, 100);
  trigger(sidebar, 'touchmove', 0, 200);
  trigger(sidebar, 'touchmove', 0, 300);
  trigger(sidebar, 'touchmove', 0, 400);
  trigger(sidebar, 'touchend', 0, 400);
  expect(fn).toHaveBeenCalledTimes(1);
  expect(onSelect).toHaveBeenCalledWith('B');
});

test('scroll and update active anchor', () => {
  const nativeRect = Element.prototype.getBoundingClientRect;
  Element.prototype.getBoundingClientRect = function () {
    const { index } = this.dataset;
    return {
      top: index ? index * 10 : 0,
    };
  };

  mockOffsetHeight(10);

  const wrapper = mount({
    template: `
      <win-index-bar :sticky="sticky">
        <win-index-anchor
          v-for="index in 4"
          :key="index"
          :index="index"
          :data-index="index - 1"
        />
      </win-index-bar>
    `,
    data() {
      return {
        sticky: false,
      };
    },
  });

  window.scrollTop = 0;
  trigger(window, 'scroll');
  expect(wrapper).toMatchSnapshot();

  wrapper.setData({ sticky: true });
  trigger(window, 'scroll');
  expect(wrapper).toMatchSnapshot();
  wrapper.vm.$destroy();

  Element.prototype.getBoundingClientRect = nativeRect;
});
