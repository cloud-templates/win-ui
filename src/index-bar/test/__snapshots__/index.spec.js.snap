// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom anchor text 1`] = `
<div class="win-index-bar">
  <div class="win-index-bar__sidebar"><span data-index="A" class="win-index-bar__index">A</span><span data-index="B" class="win-index-bar__index">B</span><span data-index="C" class="win-index-bar__index">C</span><span data-index="D" class="win-index-bar__index">D</span><span data-index="E" class="win-index-bar__index">E</span><span data-index="F" class="win-index-bar__index">F</span><span data-index="G" class="win-index-bar__index">G</span><span data-index="H" class="win-index-bar__index">H</span><span data-index="I" class="win-index-bar__index">I</span><span data-index="J" class="win-index-bar__index">J</span><span data-index="K" class="win-index-bar__index">K</span><span data-index="L" class="win-index-bar__index">L</span><span data-index="M" class="win-index-bar__index">M</span><span data-index="N" class="win-index-bar__index">N</span><span data-index="O" class="win-index-bar__index">O</span><span data-index="P" class="win-index-bar__index">P</span><span data-index="Q" class="win-index-bar__index">Q</span><span data-index="R" class="win-index-bar__index">R</span><span data-index="S" class="win-index-bar__index">S</span><span data-index="T" class="win-index-bar__index">T</span><span data-index="U" class="win-index-bar__index">U</span><span data-index="V" class="win-index-bar__index">V</span><span data-index="W" class="win-index-bar__index">W</span><span data-index="X" class="win-index-bar__index">X</span><span data-index="Y" class="win-index-bar__index">Y</span><span data-index="Z" class="win-index-bar__index">Z</span></div>
  <div>
    <div class="win-index-anchor">Title A</div>
  </div>
  <div>
    <div class="win-index-anchor">Title B</div>
  </div>
</div>
`;

exports[`scroll and update active anchor 1`] = `
<div class="win-index-bar">
  <div class="win-index-bar__sidebar"><span data-index="A" class="win-index-bar__index win-index-bar__index--active">A</span><span data-index="B" class="win-index-bar__index">B</span><span data-index="C" class="win-index-bar__index">C</span><span data-index="D" class="win-index-bar__index">D</span><span data-index="E" class="win-index-bar__index">E</span><span data-index="F" class="win-index-bar__index">F</span><span data-index="G" class="win-index-bar__index">G</span><span data-index="H" class="win-index-bar__index">H</span><span data-index="I" class="win-index-bar__index">I</span><span data-index="J" class="win-index-bar__index">J</span><span data-index="K" class="win-index-bar__index">K</span><span data-index="L" class="win-index-bar__index">L</span><span data-index="M" class="win-index-bar__index">M</span><span data-index="N" class="win-index-bar__index">N</span><span data-index="O" class="win-index-bar__index">O</span><span data-index="P" class="win-index-bar__index">P</span><span data-index="Q" class="win-index-bar__index">Q</span><span data-index="R" class="win-index-bar__index">R</span><span data-index="S" class="win-index-bar__index">S</span><span data-index="T" class="win-index-bar__index">T</span><span data-index="U" class="win-index-bar__index">U</span><span data-index="V" class="win-index-bar__index">V</span><span data-index="W" class="win-index-bar__index">W</span><span data-index="X" class="win-index-bar__index">X</span><span data-index="Y" class="win-index-bar__index">Y</span><span data-index="Z" class="win-index-bar__index">Z</span></div>
  <div data-index="0">
    <div class="win-index-anchor">1</div>
  </div>
  <div data-index="1">
    <div class="win-index-anchor">2</div>
  </div>
  <div data-index="2">
    <div class="win-index-anchor">3</div>
  </div>
  <div data-index="3">
    <div class="win-index-anchor">4</div>
  </div>
</div>
`;

exports[`scroll and update active anchor 2`] = `
<div class="win-index-bar">
  <div class="win-index-bar__sidebar"><span data-index="A" class="win-index-bar__index">A</span><span data-index="B" class="win-index-bar__index win-index-bar__index--active">B</span><span data-index="C" class="win-index-bar__index">C</span><span data-index="D" class="win-index-bar__index">D</span><span data-index="E" class="win-index-bar__index">E</span><span data-index="F" class="win-index-bar__index">F</span><span data-index="G" class="win-index-bar__index">G</span><span data-index="H" class="win-index-bar__index">H</span><span data-index="I" class="win-index-bar__index">I</span><span data-index="J" class="win-index-bar__index">J</span><span data-index="K" class="win-index-bar__index">K</span><span data-index="L" class="win-index-bar__index">L</span><span data-index="M" class="win-index-bar__index">M</span><span data-index="N" class="win-index-bar__index">N</span><span data-index="O" class="win-index-bar__index">O</span><span data-index="P" class="win-index-bar__index">P</span><span data-index="Q" class="win-index-bar__index">Q</span><span data-index="R" class="win-index-bar__index">R</span><span data-index="S" class="win-index-bar__index">S</span><span data-index="T" class="win-index-bar__index">T</span><span data-index="U" class="win-index-bar__index">U</span><span data-index="V" class="win-index-bar__index">V</span><span data-index="W" class="win-index-bar__index">W</span><span data-index="X" class="win-index-bar__index">X</span><span data-index="Y" class="win-index-bar__index">Y</span><span data-index="Z" class="win-index-bar__index">Z</span></div>
  <div data-index="0" style="height: 10px;">
    <div class="win-index-anchor win-index-anchor--sticky win-hairline--bottom" style="z-index: undefined; transform: translate3d(0, 0px, 0);">1</div>
  </div>
  <div data-index="1" style="height: 10px;">
    <div class="win-index-anchor win-index-anchor--sticky win-hairline--bottom" style="z-index: undefined; transform: translate3d(0, 10px, 0);">2</div>
  </div>
  <div data-index="2">
    <div class="win-index-anchor">3</div>
  </div>
  <div data-index="3">
    <div class="win-index-anchor">4</div>
  </div>
</div>
`;
